stix2/sources/filesystem.py

```diff
@@ -0,0 +1,201 @@
+
+        return [self.get(stix_id=stix_id, _composite_filters=_composite_filters)]
+
+    def query(self, query=None, _composite_filters=None):
+        """
+        """
+        all_data = []
+
+        if query is None:
+            query = []
+
+        # combine all query filters
+        if self.filters:
+            query.extend(self.filters.values())
+        if _composite_filters:
+            query.extend(_composite_filters)
+
+        # extract any filters that are for "type" or "id" , as we can then do
+        # filtering before reading in the STIX objects. A STIX 'type' filter
+        # can reduce the query to a single sub-directory. A STIX 'id' filter
+        # allows for the fast checking of the file names versus loading it.
+        file_filters = self._parse_file_filters(query)
+
+        # establish which subdirectories can be avoided in query
+        # by decluding as many as possible. A filter with "type" as the field
+        # means that certain STIX object types can be ruled out, and thus
+        # the corresponding subdirectories as well
+        include_paths = []
+        declude_paths = []
+        if "type" in [filter_["field"] for filter_ in file_filters]:
+            for filter_ in file_filters:
+                if filter_["field"] == "type":
+                    if filter_["op"] == '=':
+                        include_paths.append(os.path.join(self.stix_dir, filter_["value"]))
+                    elif filter_["op"] == "!=":
+                        declude_paths.append(os.path.join(self.stix_dir, filter_["value"]))
+        else:
+            # have to walk entire STIX directory
+            include_paths.append(self.stix_dir)
+
+        # if a user specifies a "type" filter like "type = <stix-object_type>",
+        # the filter is reducing the search space to single stix object types
+        # (and thus single directories). This makes such a filter more powerful
+        # than "type != <stix-object_type>" bc the latter is substracting
+        # only one type of stix object type (and thus only one directory),
+        # As such the former type of filters are given preference over the latter;
+        # i.e. if both exist in a query, that latter type will be ignored
+
+        if not include_paths:
+            # user has specified types that are not wanted (i.e. "!=")
+            # so query will look in all STIX directories that are not
+            # the specified type. Compile correct dir paths
+            for dir_ in os.listdir(self.stix_dir):
+                if os.path.abspath(dir_) not in declude_paths:
+                    include_paths.append(os.path.abspath(dir_))
+
+        # grab stix object ID as well - if present in filters, as
+        # may forgo the loading of STIX content into memory
+        if "id" in [filter_["field"] for filter_ in file_filters]:
+            for filter_ in file_filters:
+                if filter_["field"] == "id" and filter_["field"] == '=':
```

Comment:
For the first, looks like a bug to me.  For the second, absent some kind of policy stating the last "id" filter should always be used, I'd say do the efficient thing and break when you find the first.  Should "!=" be  handled on "id" filters too?